File: lua/test/roadwork_entity_test.lua
-- Roadwork entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("autobahn-api-de_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("RoadworkEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Roadwork(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = roadwork_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "roadwork." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set AUTOBAHNAPIDE_TEST_ROADWORK_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local roadwork_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.roadwork")))
    local roadwork_ref01_data = nil
    if #roadwork_ref01_data_raw > 0 then
      roadwork_ref01_data = helpers.to_map(roadwork_ref01_data_raw[1][2])
    end

    -- LIST
    local roadwork_ref01_ent = client:Roadwork(nil)
    local roadwork_ref01_match = {
      ["road_id"] = setup.idmap["road01"],
    }

    local roadwork_ref01_list_result, err = roadwork_ref01_ent:list(roadwork_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(roadwork_ref01_list_result)

    -- LOAD
    local roadwork_ref01_match_dt0 = {}
    local roadwork_ref01_data_dt0_loaded, err = roadwork_ref01_ent:load(roadwork_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(roadwork_ref01_data_dt0_loaded)

  end)
end)

function roadwork_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/roadwork/RoadworkTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read roadwork test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "roadwork01", "roadwork02", "roadwork03", "road01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("AUTOBAHNAPIDE_TEST_ROADWORK_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["AUTOBAHNAPIDE_TEST_ROADWORK_ENTID"] = idmap,
    ["AUTOBAHNAPIDE_TEST_LIVE"] = "FALSE",
    ["AUTOBAHNAPIDE_TEST_EXPLAIN"] = "FALSE",
    ["AUTOBAHNAPIDE_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["AUTOBAHNAPIDE_TEST_ROADWORK_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["AUTOBAHNAPIDE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["AUTOBAHNAPIDE_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["AUTOBAHNAPIDE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["AUTOBAHNAPIDE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/roadwork_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/autobahn-api-de-sdk/go"
	"github.com/voxgig-sdk/autobahn-api-de-sdk/go/core"

	vs "github.com/voxgig-sdk/autobahn-api-de-sdk/go/utility/struct"
)

func TestRoadworkEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Roadwork(nil)
		if ent == nil {
			t.Fatal("expected non-nil RoadworkEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"roadwork": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Roadwork(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Roadwork(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := roadworkBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "roadwork." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set AUTOBAHN_API_DE_TEST_ROADWORK_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		roadworkRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.roadwork")))
		var roadworkRef01Data map[string]any
		if len(roadworkRef01DataRaw) > 0 {
			roadworkRef01Data = core.ToMapAny(roadworkRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = roadworkRef01Data

		// LIST
		roadworkRef01Ent := client.Roadwork(nil)
		roadworkRef01Match := map[string]any{
			"road_id": setup.idmap["road01"],
		}

		roadworkRef01ListResult, err := roadworkRef01Ent.List(roadworkRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, roadworkRef01ListOk := roadworkRef01ListResult.([]any)
		if !roadworkRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", roadworkRef01ListResult)
		}

		// LOAD
		roadworkRef01MatchDt0 := map[string]any{
			"id": roadworkRef01Data["id"],
		}
		roadworkRef01DataDt0Loaded, err := roadworkRef01Ent.Load(roadworkRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		roadworkRef01DataDt0LoadResult := core.ToMapAny(entityData(roadworkRef01DataDt0Loaded))
		if roadworkRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if roadworkRef01DataDt0LoadResult["id"] != roadworkRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func roadworkBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "roadwork", "RoadworkTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read roadwork test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse roadwork test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"roadwork01", "roadwork02", "roadwork03", "road01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("AUTOBAHN_API_DE_TEST_ROADWORK_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"AUTOBAHN_API_DE_TEST_ROADWORK_ENTID": idmap,
		"AUTOBAHN_API_DE_TEST_LIVE":      "FALSE",
		"AUTOBAHN_API_DE_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["AUTOBAHN_API_DE_TEST_ROADWORK_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["AUTOBAHN_API_DE_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewAutobahnApiDeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["AUTOBAHN_API_DE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["AUTOBAHN_API_DE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
